### PR TITLE
feat: add SuperchainERC20 baseline

### LINKED ISSUE #10874: SuperchainERC20: Baseline implementation
Implement a base superchain erc20 token that can be imported and extended by other projects.

It should
- be built on top of a rock solid erc20 implementation
- have `virtual` functions so its easy to extend
- have its specific functionality unit tested

### PR DESCRIPTION
**Pull request by agusduha**
**Description**

Add a baseline standard of the [SuperchainERC20](https://github.com/ethereum-optimism/specs/pull/257).

**Tests**

Add `SuperchainERC20` unit tests

**Additional context**

We choose to go with this design to let contracts that inherits the standard be proxied or not. The downside in the current implementation is the default constructor being used in `OptimismSuperchainERC20` because of being proxied.

We would love to hear some feedback on the design to know if the ideal is a simple vanilla implementation, something like this or maybe something more focus on upgradeability.

**Metadata**

- Closes https://github.com/ethereum-optimism/optimism/issues/10874

### CODE REVIEW

File: packages/contracts-bedrock/src/L2/OptimismSuperchainERC20.sol
Comment (tynes) [inline packages/contracts-bedrock/src/L2/OptimismSuperchainERC20.sol:27]: Perhaps we could make `ISemver` part of `OptimismSuperchainERC20` to make sure we always have it in there

File: packages/contracts-bedrock/src/L2/OptimismSuperchainERC20.sol
Comment (tynes) [inline packages/contracts-bedrock/src/L2/OptimismSuperchainERC20.sol]: Do we really need this full struct now?

File: packages/contracts-bedrock/src/L2/OptimismSuperchainERC20.sol
Comment (tynes) [inline packages/contracts-bedrock/src/L2/OptimismSuperchainERC20.sol:25]: Does this need `IOptimismSuperchainERC20Extension` if its `SuperchainERC20`?

File: packages/contracts-bedrock/src/L2/OptimismSuperchainERC20.sol
Comment (agusduha) [reply packages/contracts-bedrock/src/L2/OptimismSuperchainERC20.sol]: I leave it to keep the ERC7201 implementation, but maybe we can use a slot only for the `remoteToken` and it would be enough

File: packages/contracts-bedrock/src/L2/OptimismSuperchainERC20.sol
Comment (agusduha) [reply packages/contracts-bedrock/src/L2/OptimismSuperchainERC20.sol:25]: We made a distinction between the interfaces, this one also have the `mint` and `burn` functionality used by the `L2StandardBridge`. The `ISuperchainERC20` only have the send/relay one.

**Comment by tynes** [comment]
Is this ready for review?

File: packages/contracts-bedrock/src/L2/OptimismSuperchainERC20.sol
Comment (0xng) [reply packages/contracts-bedrock/src/L2/OptimismSuperchainERC20.sol:27]: Just to double check, you meant part of `IOptimismSuperchainERC20`?

**Comment by 0xng** [comment]
@tynes It's now ready to review. We decided to make `SuperchainERC20` as minimal as possible and abstract so that contracts extending from it can choose what sort of storage to use freely.

We only have one design question, and it's around the address zero check in the `SuperchainERC20#sendERC20` function. For it to be fully compliant with the standard or a hundred percent unopionionated it should not have it. However, it seems like a check most contracts extending for it would want. We weren't sure whether to leave it or not. We removed it from `relayERC20` as it didn't make sense there if it was already checked in `sendERC20`. Any thoughts on this would be appreciated.

File: packages/contracts-bedrock/src/L2/OptimismSuperchainERC20.sol
Comment (tynes) [reply packages/contracts-bedrock/src/L2/OptimismSuperchainERC20.sol:27]: Yes exactly

**Comment by tynes** [comment]
Curious how you would feel about supporting https://github.com/ethereum-optimism/optimism/issues/11899 as part of this. The rationale is here https://github.com/ethereum-optimism/optimism/issues/11880

Since we updated the `OptimismMintableERC20` with this feature, I think it makes sense to also have it in the `OptimismSuperchainERC20`. Definitely hear your argument about being minimal, but not having this feature would create an inconsistent experience. It basically gives a superpower to the tokens

edit: I suppose this could be done in a follow up

File: packages/contracts-bedrock/test/L2/OptimismSuperchainERC20.t.sol
Comment (tynes) [inline packages/contracts-bedrock/test/L2/OptimismSuperchainERC20.t.sol]: This could be imported from `CommonErrors`

File: packages/contracts-bedrock/test/L2/SuperchainERC20.t.sol
Comment (tynes) [inline packages/contracts-bedrock/test/L2/SuperchainERC20.t.sol:46]: nit: will solc still parse this even though there is a newline?

**Comment by tynes** [comment]
Looks like the recently added interface check failed

**Comment by tynes** [comment]
> @tynes It's now ready to review. We decided to make `SuperchainERC20` as minimal as possible and abstract so that contracts extending from it can choose what sort of storage to use freely.
> 
> We only have one design question, and it's around the address zero check in the `SuperchainERC20#sendERC20` function. For it to be fully compliant with the standard or a hundred percent unopionionated it should not have it. However, it seems like a check most contracts extending for it would want. We weren't sure whether to leave it or not. We removed it from `relayERC20` as it didn't make sense there if it was already checked in `sendERC20`. Any thoughts on this would be appreciated.

Definitely on board with leaving it in there

**Comment by tynes** [comment]
Tests look great to me! Ready to merge after the checks are passing CI

File: packages/contracts-bedrock/test/L2/SuperchainERC20.t.sol
Comment (0xng) [reply packages/contracts-bedrock/test/L2/SuperchainERC20.t.sol:46]: This line is something `forge fmt` adds. I tried removing and pushing, and the linter failed.

**Comment by 0xng** [comment]
@tynes Assigned this issue to us: https://github.com/ethereum-optimism/optimism/issues/11899. Will take care of this next. Thought it was cleaner to make it in a separate PR as it touches on the `SuperchainWETH` contract as well. I also created a task for us to refine the interfaces when a standard is introduced. 

Also, as a note, for these last commits I added the main thing I felt to be untidy was to create an interface for Solady's ERC20, not due to the interface per se, but because I was not certain where to place it. I created a directory called `dependency/interface` and placed it there, but doubt it will see much use outside of this particular interface, so if you have any nit or opinion on this, let me know.  Either way, a fun thing the interface-checker showed me is that Solady's ERC20 has named returns for some functions that OZ's ERC20 doesn't.

Lastly, on the `ISemver` being part of the `IOptimismMintableERC20Extensions` I'm a bit on the fence. It is mostly because I felt it mixed two concepts, one being the additional events and functions of the `OptimismMintableERC20` and the other being versioning, but have no strong opinions here, so I can add it in a quick commit if you or anyone would like it there better.

**Review by tynes — APPROVED** [review]

**Comment by tynes** [comment]
> @tynes Assigned this issue to us: #11899. Will take care of this next. Thought it was cleaner to make it in a separate PR as it touches on the `SuperchainWETH` contract as well. I also created a task for us to refine the interfaces when a standard is introduced.
> 
> Also, as a note, for these last commits I added the main thing I felt to be untidy was to create an interface for Solady's ERC20, not due to the interface per se, but because I was not certain where to place it. I created a directory called `dependency/interface` and placed it there, but doubt it will see much use outside of this particular interface, so if you have any nit or opinion on this, let me know. Either way, a fun thing the interface-checker showed me is that Solady's ERC20 has named returns for some functions that OZ's ERC20 doesn't.
> 
> Lastly, on the `ISemver` being part of the `IOptimismMintableERC20Extensions` I'm a bit on the fence. It is mostly because I felt it mixed two concepts, one being the additional events and functions of the `OptimismMintableERC20` and the other being versioning, but have no strong opinions here, so I can add it in a quick commit if you or anyone would like it there better.

This all sounds good to me

**Pull request merged**
